aten/src/ATen/native/BatchLinearAlgebra.h

```diff
@@ -15,10 +17,73 @@ namespace at { namespace native {
 template<class scalar_t>
 void lapackEig(char jobvl, char jobvr, int n, scalar_t *a, int lda, scalar_t *wr, scalar_t *wi, scalar_t* vl, int ldvl, scalar_t *vr, int ldvr, scalar_t *work, int lwork, int *info);
 
+template<class scalar_t>
+void lapackOrgqr(int m, int n, int k, scalar_t *a, int lda, scalar_t *tau, scalar_t *work, int lwork, int *info);
+
 #endif
 
 using eig_fn = std::tuple<Tensor, Tensor> (*)(const Tensor&, bool&);
 
 DECLARE_DISPATCH(eig_fn, eig_stub);
 
+/*
+  The orgqr function allows reconstruction of an orthogonal (or unitary) matrix Q,
+  from a sequence of elementary reflectors, such as produced by the geqrf function.
+
+  Args:
+  * `self` - Tensor with the directions of the elementary reflectors below the diagonal,
+              it will be overwritten with the result
+  * `tau` - Tensor containing the magnitudes of the elementary reflectors
+  * `infos` - Tensor to store LAPACK's error codes
+  * `n_columns` - The number of columns of Q to be computed
+
+  For further details, please see the LAPACK documentation for ORGQR and UNGQR.
+*/
+template <typename scalar_t>
+inline void apply_orgqr(Tensor& self, const Tensor& tau, Tensor& infos, int64_t n_columns) {
+#ifndef USE_LAPACK
+  TORCH_CHECK(false, "Calling torch.orgqr on a CPU tensor requires compiling ",
+    "PyTorch with LAPACK. Please use PyTorch built with LAPACK support.");
+#else
+  using value_t = typename c10::scalar_value_type<scalar_t>::type;
+  auto self_data = self.data_ptr<scalar_t>();
+  auto tau_data = tau.data_ptr<scalar_t>();
+  auto infos_data = infos.data_ptr<int>();
+  auto self_matrix_stride = matrixStride(self);
+  auto tau_stride = tau.size(-1);
+  auto batch_size = batchCount(self);
+  auto m = self.size(-2);
+  auto k = tau.size(-1);
+  auto lda = std::max<int64_t>(1, m);
+
+  // LAPACK's requirement
+  TORCH_INTERNAL_ASSERT(m >= n_columns);
+  TORCH_INTERNAL_ASSERT(n_columns >= k);
+
+  // Run once, first to get the optimum work size.
+  // Since we deal with batches of matrices with the same dimensions, doing this outside
+  // the loop saves (batch_size - 1) workspace queries which would provide the same result
+  // and (batch_size - 1) calls to allocate and deallocate workspace using at::empty()
+  int lwork = -1;
+  scalar_t wkopt;
+  lapackOrgqr<scalar_t>(m, n_columns, k, self_data, lda, tau_data, &wkopt, lwork, &infos_data[0]);
+  lwork = static_cast<int>(real_impl<scalar_t, value_t>(wkopt));
```

Comment:
I don't like this error `Argument 8 has illegal value`. The 8-th argument is `lwork`, it is an integer for which we should obtain the value from LAPACK.
> The size of the work array (lwork≥n).

MKL doc link: https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-computational-routines/orthogonal-factorizations-lapack-computational-routines/orgqr.html

Reply:
@mruberry ci-all is all green (except [this](https://app.circleci.com/pipelines/github/pytorch/pytorch/261921/workflows/6ee88bb2-78c5-4545-bd9e-fae39b278758/jobs/10265138) which seems unrelated).
Could you somehow check what BLAS/LAPACK implementation PyTorch in that failing build is linked to? Is it MKL?`torch.__config__.show()`

Reply:
The versioned used internally seems to be much older than the version in CI.

Could it be the change in how the function is called, using m instead of lda? I'll try to debug internally, too, to get a better sense for what's going on. @ngimel points out that torch.linalg.qr must be relying on this same function, so it's surprising we haven't seen this issue previously.